processor/src/host/mod.rs

```diff
@@ -91,20 +119,30 @@ pub trait AsyncHost: BaseHost {
     // REQUIRED METHODS
     // --------------------------------------------------------------------------------------------
 
-    // Note: we don't use the `async` keyword in both of these methods, since we need to specify the
-    // `+ Send` bound to the returned Future, and `async` doesn't allow us to do that.
+    // Note: we don't use the `async` keyword in this method, since we need to specify the `+ Send`
+    // bound to the returned Future, and `async` doesn't allow us to do that.
 
-    /// Returns MAST forest corresponding to the specified digest, or None if the MAST forest for
-    /// this digest could not be found in this [AsyncHost].
-    fn get_mast_forest(
-        &self,
-        node_digest: &Word,
-    ) -> impl Future<Output = Option<Arc<MastForest>>> + Send;
-
-    /// Handles the event emitted from the VM.
+    /// Handles the event emitted from the VM and provides advice mutations to be applied to
+    /// the advice provider.
     fn on_event(
         &mut self,
-        process: &mut ProcessState<'_>,
+        process: &ProcessState<'_>,
         event_id: u32,
-    ) -> impl Future<Output = Result<(), EventError>> + Send;
+    ) -> impl FutureMaybeSend<Result<Vec<AdviceMutation>, EventError>>;
 }
+
+/// Alias for a `Future`
+///
+/// If feature `std` is enabled, we add `Send` to the required bounds, otherwise we do not. This
+/// impacts usability with a multithreaded executor.
+#[cfg(target_family = "wasm")]
+pub trait FutureMaybeSend<O>: Future<Output = O> {}
+
+#[cfg(target_family = "wasm")]
+impl<T, O> FutureMaybeSend<O> for T where T: Future<Output = O> {}
+
+#[cfg(not(target_family = "wasm"))]
+pub trait FutureMaybeSend<O>: Future<Output = O> + Send {}
```

Comment:
there's no docs when `not(target_family = "wasm"))`, so I think the standard generated docs wouldn't have a docstring for this trait? Therefore we should copy over the docstring here